owlkettle/playground.nim

```diff
@@ -123,165 +126,142 @@ method view (state: DateDialogState): Widget =
           proc select(date: DateTime) =
             state.date = date
 
-proc toFormField(state: auto, fieldName: static string, typ: typedesc[DateTime]): Widget =
-  ## Provides a form field for DateTime
+proc toFormField(state: Viewable, field: ptr DateTime, fieldName: string): Widget =
+  ## Provides a form field for DateTimes
   return gui:
     ActionRow:
       title = fieldName
-      subtitle =  $state.getField(fieldName).inZone(local())
+      subtitle =  $(field[]).inZone(local())
       Button {.addSuffix.}:
         icon = "x-office-calendar-symbolic"
         text = "Select"
         proc clicked() =
           let (res, dialogState) = state.open(gui(DateDialog()))
           if res.kind == DialogAccept:
-            state.getField(fieldName) = DateDialogState(dialogState).date
+            field[] = DateDialogState(dialogState).date
+
+
+proc toFormField(state: Viewable, field: ptr[distinct], fieldName: string): Widget =
+  ## Provides a form field for any distinct type. 
+  ## The form field provided is the same that the base-type of the distinct type would have.
+  let baseField = field[].distinctBase.addr
```

Comment:
For reference: I double checked that this does not change the actual pointer.
The memory pointed to by baseField and field is the same, just the type that is assumed behind them is different.

For reference, checkout this example:
```nim
import std/typetraits

type X = distinct string

proc debug(x: ptr auto) = 
  echo cast[uint64](x), " - ", x[].type, " - ", x[].distinctBase

let x: X = "lala".X
let y = x.addr
debug y

let z = y[].distinctBase.addr
debug z
```